shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -308,31 +308,39 @@ - (void)removeSplashScreenView:(dispatch_block_t _Nullable)onComplete {
 }
 
 - (void)installFirstFrameCallback {
-  auto weak_platform_view = [_engine.get() platformView];
+  fml::WeakPtr<flutter::PlatformViewIOS> weak_platform_view = [_engine.get() platformView];
   if (!weak_platform_view) {
     return;
   }
-  __unsafe_unretained auto weak_flutter_view_controller = self;
+
   // This is on the platform thread.
-  weak_platform_view->SetNextFrameCallback([weak_platform_view, weak_flutter_view_controller,
-                                            task_runner = [_engine.get() platformTaskRunner]]() {
-    // This is on the GPU thread.
-    task_runner->PostTask([weak_platform_view, weak_flutter_view_controller]() {
-      // We check if the weak platform view is alive. If it is alive, then the view controller
-      // also has to be alive since the view controller owns the platform view via the shell
-      // association. Thus, we are not convinced that the unsafe unretained weak object is in
-      // fact alive.
-      if (weak_platform_view) {
-        if (weak_flutter_view_controller->_splashScreenView) {
-          [weak_flutter_view_controller removeSplashScreenView:^{
-            [weak_flutter_view_controller callViewRenderedCallback];
-          }];
-        } else {
-          [weak_flutter_view_controller callViewRenderedCallback];
-        }
-      }
-    });
-  });
+  weak_platform_view->SetNextFrameCallback(
+      [weak_platform_view, platform_task_runner = [_engine.get() platformTaskRunner],
+       gpu_task_runner = [_engine.get() gpuTaskRunner]]() {
+        FML_DCHECK(gpu_task_runner->RunsTasksOnCurrentThread());
+        platform_task_runner->PostTask([weak_platform_view]() {
+          // The weak platform view will have a weak pointer to self, if it is
+          // still alive. We can check if that weak pointer is still valid to
+          // know if we're still alive.
+          if (weak_platform_view) {
+            fml::WeakPtr<FlutterViewController> weak_self =
+                weak_platform_view->GetOwnerViewController();
+            if (!weak_self) {
```

Comment:
The check should be on the strong pointer, after converting weak to strong. (It only actually matters when things could be released on another thread, but the convention is a fairly common one in this type of ObjC code and it's easier for readers who are familiar with it if you follow the convention.)

Reply:
fml::WeakPtr has an operator defined for this, and this is more typically how it's used throughout the codebase AFAICT.

Reply:
This isn't about fml::WeakPtr, it's about following standard ObjC patterns (even though the specific types are unusual). The standard, recommended approach for doing any non-trivial logic (i.e., more than a single call) on a weak pointer in a callback block is:
1) convert weak pointer to strong point
2) check that the strong pointer is non-nil; if it is don't do anything else
3) do whatever logic, using only the strong pointer

Reversing steps 1 and 2, as you have here, creates a race condition in some situations, and has no advantages.

Reply:
I've actually simplified this a bit further - but still checking before retaining.  Does that look good, or do you really want me to call get and retain on what's potentially a null pointer?

Reply:
> do you really want me to call get and retain on what's potentially a null pointer

Yes I do, for the reasons described in my two previous comments. The last revisions didn't change the pattern.

(And to be pedantic, you're not calling `get()` on a null pointer, which sounds like a, but calling `get()` on a perfectly valid object to return a pointer which may be nil. And retaining nil is a well-defined no-op.)

Reply:
TIL.  Updated.